### ISSUE TITLE
PHP 7.3 and PCRE2

### ISSUE THREAD

**Issue by jkbzh** [squizlabs/PHP_CodeSniffer#2741]
Hi,

PHP 7.3  [migrated from PCRE to PCRE2 ](https://www.php.net/manual/en/migration73.other-changes.php#migration73.other-changes.pcre).  PCRE2 is [stricter](https://stackoverflow.com/questions/57111993/php7-3-regex-something-went-wrongurl) and requires escaping of hiphens in regular expressions. This is not detected by PHPCS. If this makes sense to you, could it be possible to add a test for hiphens and other characters that need to be escaped so that the following sample would raise a REGEXP warning when running phpcs?

```
<?php

$tag="j";

if (preg_match("/^[\w-:]+$/", $tag)) {
    print "match";
} else {
    print "no match";
}

```
PHP 7.3 on the above code gives: 
`Warning: preg_match(): Compilation failed: invalid range in character class at offset 4 in /usr/local/src/test.php3 on line 5`

Thanks!

**Comment by gsherwood** [comment]
Just to confirm - you are suggesting a new sniff to try and detect invalid regular expressions?

**Comment by jrfnl** [comment]
@jkbzh That would be a sniff I'd happily add to [PHPCompatibility](https://github.com/PHPCompatibility/PHPCompatibility/), though won't be easy to sniff for unless the regex is a string literal within a regex function call. If it's a variable/constant etc, we won't have access to the runtime value of the variable, so we're out of luck.

To write it, I'd need quite some code samples though, so any additional code samples you can think of would be helpful.

Oh and [opening a feature request](https://github.com/PHPCompatibility/PHPCompatibility/issues/new) about it in the PHPCompatibility repo would be appreciated.

**Comment by austinpray** [comment]
Here's some code samples!

## [broken](https://3v4l.org/U8ET5)

```php
<?php

$path = '/carrot/v1/site/43050/analytics';
$route = '/carrot/v2/sites/(?P<blog_id>\d+)/migrations/(?P<migration>[\w-_]+)';

var_dump(preg_match( '@^' . $route . '$@i', $path, $matches ));
var_dump($matches);
```

## [fixed](https://3v4l.org/hpjhD)

```php
<?php

$path = '/carrot/v1/site/43050/analytics';
$route = '/carrot/v2/sites/(?P<blog_id>\d+)/migrations/(?P<migration>[\w\-_]+)';

var_dump(preg_match( '@^' . $route . '$@i', $path, $matches ));
var_dump($matches);
```

**Comment by jrfnl** [comment]
I've opened an issue in the PHPCompatibility repository about this: https://github.com/PHPCompatibility/PHPCompatibility/issues/965

**Comment by jrfnl** [comment]
Closing as this type of sniff is typically not accepted in PHPCS itself and their is an external standard which could address this.
